### ISSUE TITLE
trace merge error for inconsistent sampling rates

### ISSUE THREAD

**Issue by bch0w** [adjtomo/pysep#80]
Occasionally a trace downloaded from a data center will not have the same expected sampling rate. E.g., IU.MBAR.00.LHZ downloaded from IRIS has segmented data with sampling rates which look like floating point rounding errors

```python
print(st_edit_select)
3 Trace(s) in Stream:
II.MBAR.00.LHZ | 2004-12-25T12:58:53.800400Z - 2004-12-25T13:54:43.814377Z | 1.0 Hz, 3351 samples
II.MBAR.00.LHZ | 2004-12-25T14:02:55.749800Z - 2004-12-26T01:26:54.930813Z | 1.0 Hz, 41040 samples
II.MBAR.00.LHZ | 2004-12-26T01:36:31.748600Z - 2004-12-26T13:13:16.928008Z | 1.0 Hz, 41806 samples

for tr in st_edit_select: 
    print(tr.stats.sampling_rate)
0.9999958276748657
0.9999955892562866
0.9999957084655762
```

This will cause ObsPy to crash when calling the merge() function.

```bash
/opt/miniconda3/envs/pysep/lib/python3.11/site-packages/obspy/core/stream.py:3037: UserWarning: Incompatible traces (sampling_rate, dtype, ...) with same id detected. Doing nothing.
  warnings.warn(msg)
Traceback (most recent call last):
  File "/Users/chow/Work/scratch/sumatra_test.py", line 27, in <module>
    ps.run()
  File "/Users/chow/Repos/pysep/pysep/pysep.py", line 1558, in run
    self.st = self.preprocess()
              ^^^^^^^^^^^^^^^^^
  File "/Users/chow/Repos/pysep/pysep/pysep.py", line 1107, in preprocess
    st_out = merge_and_trim_start_end_times(st_out)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/chow/Repos/pysep/pysep/utils/process.py", line 240, in merge_and_trim_start_end_times
    st_edit_select.merge()  # combining like trace IDs
    ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/pysep/lib/python3.11/site-packages/obspy/core/stream.py", line 2037, in merge
    self._merge_checks()
  File "/opt/miniconda3/envs/pysep/lib/python3.11/site-packages/obspy/core/stream.py", line 1975, in _merge_checks
    raise Exception(msg)
Exception: Can't merge traces with same ids but differing sampling rates!
```

The fix this, we can run the resample function **before** doing the trace merge and trim, to ensure that we don't hit this

**Comment by aakash10gupta** [comment]
@bch0w,

It looks like some seismograms are still getting rejected with the following error message -

`[2023-02-26 01:28:37] - pysep - INFO: II.PALK.00.LHZ has data gaps, filling with: 0`
`[2023-02-26 01:28:37] - pysep - WARNING: II.PALK.00.LHZ merge error: Can't merge traces with same ids but differing sampling rates!`

Even though there is no error and the rest of the process goes on smoothly, is it intended to reject such waveforms with sampling frequency discrepancies. Here ([trace_merge_test.txt](https://github.com/adjtomo/pysep/files/10832694/trace_merge_test.txt)) is a python script to emulate the error.

**Comment by aakash10gupta** [comment]
Ok, I am noticing these waveforms are only accepted when a resampling is explicitly requested, like `resample_freq = 1`
The default behavior which is `resample_freq = None` will reject these waveforms.

It might be useful in some cases to have data which is originally available in two (or more) patches for the given time window, with sampling frequencies like `1Hz` and `0.99995Hz` for each patch respectively, merged into one stream with a uniform sampling frequency.

Is the idea that the user explicitly specifies a resampling frequency if the data is expected to have a `~1Hz` sampling frequency? Can such a resampling request modify data which did not need resampling in the first place?

**Comment by carltape** [comment]
From what I recall, resample_freq was offered because some of the moment tensor codes required all seismograms to have the same sample rate. So if your original seismograms had 20, 40, 50 sps, you would set resample_freq = 20 sps (or whatever you needed) to ensure this. Not sure if this history is helpful here.

**Comment by bch0w** [comment]
> Ok, I am noticing these waveforms are only accepted when a resampling is explicitly requested, like `resample_freq = 1` The default behavior which is `resample_freq = None` will reject these waveforms.

Good catch, should have mentioned that PySEP currently requires a known sampling rate to resample at, otherwise you will still get the inconsistent sampling rate merge problem.

> It might be useful in some cases to have data which is originally available in two (or more) patches for the given time window, with sampling frequencies like `1Hz` and `0.99995Hz` for each patch respectively, merged into one stream with a uniform sampling frequency.

So yes, if you set `resample_freq = 1` then the two patches with inconsistent sampling rates will be first resampled to 1 Hz and then merged together.

> Is the idea that the user explicitly specifies a resampling frequency if the data is expected to have a `~1Hz` sampling frequency? Can such a resampling request modify data which did not need resampling in the first place?

Currently yes, as PySEP does not implicitly know what sampling rate it should be getting, so both 1 Hz and 0.99995 Hz could be valid. The idea was to let the User dictate what sampling rate they wanted, otherwise you get what you get from the data center which we are seeing is not always uniform.

We could try to find the mode or mean value of sampling rate and force that value on all other traces, but as we see from the criteria to remove for insufficient trace length, which does the same thing but for start and end times, this doesn't really work in all cases.

**Comment by aakash10gupta** [comment]
Thanks @bch0w.

> Might be worth it to have a 'why are my stations being removed' section, with these common errors and solutions

How about having pysep's parameter defaults set so that all available data is accepted? This will force users to consciously toggle/set parameters to discard data not fulfilling certain criteria, and hence users will be more aware of the data selection procedure.

**Comment by bch0w** [comment]
I would argue that it's better to have PySEP be the most restrictive by default, in the case that someone comes in blind and just wants data for a moment tensor inversion, returning inconsistent data will likely cause downstream problems that they may or may not be aware of.

In contrast, those that want less restrictions should be responsible for turning things off selectively and understanding the consequences of those actions.

Bottom line, I think this all boils down to PySEP's main goal: which is to provide curated, consistent data for inversions. If you want all data available, then it's easier to directly use ObsPy or FDSN.

**Comment by aakash10gupta** [comment]
That's a good point. In that case the argument `remove_clipped = False` would need a toggled default value. Other arguments look set to be most restrictive at the moment.
